### PR TITLE
Proposal: Allow events to initiate builds via new Cloud Event Listener CRD

### LINKED ISSUE #577: Allow events to initiate Builds
The eventing controller seems to be focused on routing events to Elafros apps via routes.

If it were also possible to initiate builds in response to events of all of the types supported by eventing, we could make a big stride toward a general-purpose K8s-native CI platform, supporting builds in response to webhooks, or on a schedule, or in response to PubSub events, etc. If builds were able to automatically trigger deployments, we'd get CD as well.

Sharing these concepts and triggering infrastructure could help reduce new code that would have to be written and maintained, and new concepts that would have to be understood by users.

### PR DESCRIPTION
**Pull request by iancoffey**
Fixes #577 

Depends:

- [x] https://github.com/knative/eventing-sources/pull/309

## Proposed Changes
This is a functional proposal which introduces a means for knative eventing to directly trigger builds upon receipt of specific events. The build listener accepts cloud events and produces builds for them. Currently this proposal only accepts Github `check_suite` events, but any event can theoretically produce any build.

An overview is provided in the [readme](https://github.com/knative/build/blob/86a73694aa27bbaf3a82d859541b055fb49cd95b/cmd/cloudevent-listener/README.md).

* Propose a new CRD CloudEventsListener (CEL) to provide a simple and configurable interface between Eventing and Builds.
* The CEL itself is a small server that accepts Cloudevents and creates a Build object, which is configured via the CRD definition. The raw Build spec is mounted into the listener via a secret, to protect whatever sensitive things that may be inside.
* The CEL is meant to be deployed to listen for a specific build pipeline and multiple CEL listeners can be deployed in a single cluster to handle multiple events/builds/pipelines.
* Each CEL is provisioned with a desired Build in mind as well as a specific event and is populated with specific commit information from the event.
* Once a build is submitted, it behaves and is configured no differently than another other Build object (because it is just a normal Build object).
* Only github `check_suite` events are supported by this proposal but the design is meant to provide an interface between any compatible event from any source and any Build configuration.

*This PR is very much a proposal and a wip, created to generate feedback*

Cloud Events Listener Flow

**Release Note**

```release-note

```

### CODE REVIEW

File: config/500-cloudeventslistener.yaml [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline config/500-cloudeventslistener.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

File: config/build-config.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Copyright 2018 Knative Authors LLC
```
Comment (mattmoor-sockpuppet) [inline config/build-config.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: config/cloudeventlistener.yaml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# Copyright 2018 Knative Authors LLC
```
Comment (mattmoor-sockpuppet) [inline config/cloudeventlistener.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go [hunk 5]
```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go [hunk 6]
```diff
@@ -0,0 +1,137 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go [hunk 7]
```diff
@@ -0,0 +1,86 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2018 The Knative Authors
```
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: config/500-cloudeventslistener.yaml (hunk 1 above)
Comment (mattmoor-sockpuppet) [inline config/500-cloudeventslistener.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

File: config/build-config.yaml (hunk 2 above)
Comment (mattmoor-sockpuppet) [inline config/build-config.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: config/cloudeventlistener.yaml (hunk 3 above)
Comment (mattmoor-sockpuppet) [inline config/cloudeventlistener.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go (hunk 4 above)
Comment (mattmoor-sockpuppet) [inline pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go (hunk 5 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go (hunk 6 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go (hunk 5 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go (hunk 6 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go (hunk 5 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go (hunk 6 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/clientset/versioned/typed/build/v1alpha1/fake/fake_cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: config/500-cloudeventslistener.yaml (hunk 1 above)
Comment (mattmoor-sockpuppet) [inline config/500-cloudeventslistener.yaml]:
```suggestion
# Copyright 2019 The Knative Authors
```

File: config/build-config.yaml (hunk 2 above)
Comment (mattmoor-sockpuppet) [inline config/build-config.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: config/cloudeventlistener.yaml (hunk 3 above)
Comment (mattmoor-sockpuppet) [inline config/cloudeventlistener.yaml]:
```suggestion
# Copyright 2019 Knative Authors LLC
```

File: pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go (hunk 4 above)
Comment (mattmoor-sockpuppet) [inline pkg/apis/build/v1alpha1/cloud_events_listener_types_test.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: cmd/cloudevent-listener/README.md [hunk 9]
```diff
@@ -0,0 +1,11 @@
+# Cloud Events Listener
+
+The cloud events listener proposal defines a process by which Knative Eventing can directly trigger Builds.
+
+To do this, an optional CRD `CloudEventsListeners` is provided, which provides a method for Builds to act as an Eventing sink and accept CloudEvents. This makes it possible for any event to trigger any build, providing a lightweight pipeline for common tasks like Image Building.
+
+Inside the CEL spec, a build is defined. Once applied, this new CRD will deploy a small listener which will listen for a specific cloud event from a specific source. Once that has been recieved, the listener populates the SHA we wish to build inside the custom Build that was provided in the Spec.
```
Comment (mattmoor-sockpuppet) [inline cmd/cloudevent-listener/README.md]:
```suggestion
Inside the CEL spec, a build is defined. Once applied, this new CRD will deploy a small listener which will listen for a specific cloud event from a specific source. Once that has been received, the listener populates the SHA we wish to build inside the custom Build that was provided in the Spec.
```

**Comment by ImJasonH** [comment]
Thanks for this PR! Sorry for the delay in responding. This looks great! There's just one problem :confused: :

At this time the future of Knative Build is a bit murky, and we're working through how it should interact with things like Eventing and Serving, and what kinds of usability benefits it provides. The main tension is that the "Build" concept as currently defined might be overly restrictive for real-world build-test-deploy use cases -- steps only execute in order within the same Pod, and are currently tied to Serving deployments, since that was their original use case.

However, the [Tekton Pipelines](https://github.com/tektoncd/pipeline) (née Knative Pipelines) project is working toward a full CI/CD platform on Kubernetes, with support for Knative deployments, as well as deployments to anywhere, and non-deployment steps like tests and vulnerability scans. Pipelines aims to support complex workflows, retries, results and artifacts, everything.

Tekton is [currently actively designing](https://docs.google.com/document/d/1vxFEovcLxDbypYMTAS1r6t8KCL4ltGmck266tbS1mTw/edit) how it will integrate events like this to automatically (i.e., _continuously_) trigger pipeline executions. I think they'd be very interested to learn about this proposal, and if you feel like discussing/demoing this at the weekly working group meeting, it's [every Tuesday at 9am PST](https://calendar.google.com/calendar/r/eventedit/copy/amZyNTljOWpkZWdibmpsY3JlazNodDU5NWdfMjAxOTAzMjZUMTYwMDAwWiBnb29nbGUuY29tX2Qzb3Zjdm8xcDMyMTloOTg5NTczdjk4Zm5zQGc/amFzb25oYWxsQGdvb2dsZS5jb20?scp=ALL&sf=true) -- just add yourself to [the agenda](https://docs.google.com/document/d/1rPR7m1Oj0ip3bpd_bcS1sjZyPgGi_g9asF5YrExeESc/edit).

Let me know if you have any questions about Tekton, or reach out [#build-pipeline on Slack](https://knative.slack.com/messages/build-pipeline) and we'd be happy to answer any questions.

Or, if you'd prefer to find a fit for this inside Knative Build, I'd love to chat about that too, we can do that here, or [on Slack](https://knative.slack.com/messages/build-crd), or at the next [Knative Build working group meeting](https://www.knative.dev/contributing/working-groups/#build), which is next week.

This is truly very exciting work and we'd love your input. :+1:

**Comment by iancoffey** [comment]
@ImJasonH Wow, that is a lot of great information! Thanks much. 

I cant actually see any of the Tekton docs you linked yet or the calendar but I have requested access, so Im very exited to read through that discussion. Also, I will def attend the next working group meeting Tuesday.

Unfortunately, I cant seem to sign up for a knative slack account since I lack the appropriate company email (I work for Salesforce), which is a major bummer. Can I be invited to Slack perhaps?

As for where this logic lives, that question is one of the many that originally inspired this proposal - it started as a standalone project, but seemed a decent fit here given the context in https://github.com/knative/build/issues/577. That said, one nice thing about this approach is that it is portable and simple to move to another project, or become its own. The more I think over this approach however, the more I realize that I do really like this general approach. It is just so simple to configure and the isolation/security of data seems a strong point. Also, other things could be made to control the listeners for a truly large scale deployment scenario.

Thanks again and hopefully we can continue to chat on slack and those documents soon when I get access 😄

**Comment by evankanderson** [comment]
https://github.com/knative/docs/blob/master/contributing/SLACK-GUIDELINES.md has info on joining slack; the actual signup is at https://slack.knative.dev/

I'll take a look at this piece later today, but I think this may be a "missing link" between source repo changes and triggering CD, and it seems like a useful component that could be included / vendored in multiple locations as needed.

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: cmd/cloudevent-listener/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+
+This service could pretty easily be extended to allow it to be able to emit a CloudEvent of its own to whatever endpoint is defined.
+
+# Minikube instructions
+
+To get this bootstrapped locally:
+
+
+* Get the `ko` command: `go get -u github.com/google/ko/cmd/ko`
+* Load your docker enviroment vars: `eval $(minikube docker-env)`
```
Comment (mattmoor-sockpuppet) [inline cmd/cloudevent-listener/README.md]:
```suggestion
* Load your docker environment vars: `eval $(minikube docker-env)`
```

File: pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go (hunk 7 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/informers/externalversions/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

File: pkg/client/listers/build/v1alpha1/cloudeventslistener.go (hunk 8 above)
Comment (mattmoor-sockpuppet) [inline pkg/client/listers/build/v1alpha1/cloudeventslistener.go]:
```suggestion
Copyright 2019 The Knative Authors
```

**Comment by iancoffey** [comment]
At this point, I can actually end to end test this locally with minikube and example cloudevents loaded with a github check_suite payloads. It actually works! Builds are made, everything seems set as I hoped it would be. That means that if/when I get this deployed somewhere other than locally, I should have a full end to end pipeline for simple eventing->build integration. And its tiny. Which is pretty neat.

Screen Shot 2019-04-09 at 12 42 34 PM

I added a bunch more documentation on this proposal [here](https://github.com/knative/build/blob/2097bae7db58491640b8530dfc865faad782e3de/cmd/cloudevent-listener/README.md) in the cmd dir.

So this experiment has succeeded I guess, it can work end to end. 

What is the feeling for bringing this or something similar in and if its possible, what needs to happen before that can happen? I really am enjoying working with this crd, it really opens a lot of possibilities for my work and I would suppose many others as well.

@evankanderson @ImJasonH

**Comment by iancoffey** [comment]
Also, I found [this tekton proposal](https://docs.google.com/document/d/1vxFEovcLxDbypYMTAS1r6t8KCL4ltGmck266tbS1mTw/edit?disco=AAAACu6Ex1I&ts=5c90ad27) and the crd in this PR might be a partial fit there, and could maybe be modified to just handle the event differently for whatever tekton expects.

**Comment by iancoffey** [comment]
I am going to close this in favor of my Tekton PR -> https://github.com/tektoncd/pipeline/pull/783 - but if someone finds this useful please do re-open.

Thanks @ImJasonH and @evankanderson for the help!

**Pull request closed without merging**